### ISSUE TITLE
[BUG] Environment variables not interpreted properly when extending compose file

### ISSUE THREAD

**Issue by Batleram** [docker/compose#12001]
### Description

I have a root docker-compose.yml file that contains some mounted volumes use environment variables to resolve their path
ex: $HOME/.Xauthority

When I extend that docker-compose.yml file and try to run the extended service, I get an error about an undefined volume:
`service "test_service_2" refers to undefined volume /home/user/.Xauthority: invalid compose project`

This error has appeared in the newest version of docker-compose-plugin. 
When I downgrade `docker-compose-plugin` to `2.28.1-1~ubuntu.22.04~jammy`, everything works properly.

I am able to fix this issue by adding a leading `/` to the mounted volume, to force it to be interpreted as a mount and not a local volume, but this is still a bug that needs to be fixed.

### Steps To Reproduce

Create docker-compose2.yml
```yml
services:
  test_service:
    image: busybox
    volumes:
      - "$HOME/.Xauthority:/root/.Xauthority:ro"
```
Create docker-compose.yml
```yml
services:
  test_service_2:
    extends:
      file: ./docker-compose2.yml
      service: test_service
```

Then run:
```bash
docker compose run -it test_service_2
```

### Compose Version

```Text
Docker compose version:
`Docker Compose version v2.29.0`

Docker-compose version:
Not installed
```

### Docker Environment

```Text
docker info

Client: Docker Engine - Community
 Version:    27.1.0
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.16.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.29.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 170
  Running: 3
  Paused: 0
  Stopped: 167
 Images: 45
 Server Version: 27.1.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41
 runc version: v1.1.13-0-g58aa920
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.5.0-44-generic
 Operating System: Ubuntu 22.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.712GiB
 Name: asdf-ternary
 ID: da429239-ba85-4f64-9dec-2ddc57bdcea7
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by idsulik** [comment]
@ndeloof hi! I researched it and the bug's reason is this changes https://github.com/compose-spec/compose-go/pull/651 , 
`github.com/compose-spec/compose-go/v2/format/volume.go:148` - here it receives not interpolated `volume.Source` and it sets it as `volume` type, not `bind`.  I wanted to fix it, but I don't know why did you move interpolation from `loadYamlFile` into `loadYamlModel`.  
One way to fix it is to put interpolation after this line:
github.com/compose-spec/compose-go/v2/loader/extends.go:176
```
source, processor, err := loadYamlFile(ctx, types.ConfigFile{Filename: local},
			extendsOpts, relworkingdir, nil, ct, map[string]any{}, nil)
```

**Comment by aleDVirgo** [comment]
Having the same issue here: extending a compose local file is causing the invalid volume error... and I was already using absolute paths. Reverted back to 2.28.1 fixed the issue

**Comment by johnraz** [comment]
The same applies to the `pull_policy` attribute:

`docker-compose.bug.yml`

```
services:
  mailhog:
    image: mailhog/mailhog:latest
    pull_policy: ${IMAGE_PULL_POLICY}
```

This gives:

```
IMAGE_PULL_POLICY=always docker compose -f docker-compose.bug.yml config
validating <redacted>/docker-compose.bug.yml: services.mailhog.pull_policy services.mailhog.pull_policy must be one of the following: "always", "never", "if_not_present", "build", "missing"
```

Reverting to 2.28.1-1 fixes the problem.

**Comment by t3chn0m4g3** [comment]
Can confirm as well. Downgrade to 2.28 resolves it...

Run `sudo apt-cache madison docker-compose-plugin` and it will spit out all the available repo versions:
```
docker-compose-plugin | 2.29.0-1~debian.12~bookworm | https://download.docker.com/linux/debian bookworm/stable arm64 Packages
docker-compose-plugin | 2.28.1-1~debian.12~bookworm | https://download.docker.com/linux/debian bookworm/stable arm64 Packages
[...]
```

Then ...
`sudo apt-get install docker-compose-plugin=2.28.1-1~debian.12~bookworm`

**Comment by marcu87** [comment]
A workaround I found is to start the path with a `/` to ensure it is treated as absolute. For example:

```
services:
  test_service:
    image: busybox
    volumes:
      - "/$HOME/.Xauthority:/root/.Xauthority:ro"
```

This approach isn't ideal, but it will work until a more permanent fix is available.

The issue appears to occur when a path starts with a variable; Docker Compose interprets such paths as relative instead of absolute.

**Comment by ndeloof** [comment]
fixed in v2.29.1

**Comment by johnraz** [comment]
@ndeloof when is it usually pushed to ubuntu registries ?

**Comment by ndeloof** [comment]
@johnraz https://github.com/docker/docker-ce-packaging/pull/1041

**Comment by ap-wtioit** [comment]
Would have been nice to have a proper entry in fixes for 2.29.1 https://github.com/docker/compose/releases/tag/v2.29.1 for this, instead of just "bump compose-go v2.1.5" in Internal.
